### DISCUSSION TITLE
Is it possible to install PWAsForFirefox without Admin Privileges?

### DISCUSSION THREAD

**Discussion by hugorossi** [filips123/PWAsForFirefox#111] [Q&A] [answered]
I'm working on a PC without admin privileges and would still like to install PWAsForFirefox. Is this possible?

Thanks for your reply.

Hugo

**Comment by filips123** [comment]
Probably, but you will have to change source code a bit and compile and install it yourself, because PWAsForFirefox expects some files (executables, native messaging manifest and UserChrome modifications) to be installed in some pre-defined global directory (by default `C:\Program Files\FirefoxPWA` on Windows).

1. Set `FFPWA_EXECUTABLES` and `FFPWA_SYSDATA` build-time environment variables ([source](https://github.com/filips123/PWAsForFirefox/blob/main/native/src/directories.rs#L23-L70)) to some writable directory and build the project.
2. Copy `firefoxpwa.exe` and `firefoxpwa-connector.exe` to a directory specified in `FFPWA_EXECUTABLES`, and [`userchrome`](https://github.com/filips123/PWAsForFirefox/tree/main/native/userchrome) to a directory in `FFPWA_SYSDATA`.
3. Copy [manifest for your OS](https://github.com/filips123/PWAsForFirefox/tree/main/native/manifests) to [the correct location](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_manifests#manifest_location) and edit it so it points to `firefoxpwa-connector.exe`. In case of Windows, you can just copy manifest anywhere you want (but make sure it's named `firefoxpwa.json`), then create registry key `HKEY_CURRENT_USER\SOFTWARE\Mozilla\NativeMessagingHosts\firefoxpwa` that contains a path to the manifest. You will probably have to do this through command line as you cannot access `regedit` without admin privileges.

I know it's a bit complicated and I haven't tested it, but I think it should work.

**Reply by hugorossi** [reply]
Thanks for your detailed reply. Although I must say this is way beyond my ability to do any such thing. Thank you anyway, I appreciate it very much that you took the time to write this.

**Comment by filips123** [comment, marked as answer]
Since I first answered this, support for per-user installs has been added to the MSI installer.

However, if you choose the per-user option, you must remember to always use it when installing updates, otherwise it will cause problems. Also, it might not work correctly when installing it using winget or chocolately.
